### ISSUE TITLE
AV1 encode not generating any B-frames? 

### ISSUE THREAD

**Issue by IsmetBesic** [rigaya/NVEnc#438]
Greetings! 

Was toying around with the encoder when I noticed that B-frames never appeared to be generated when using the -c av1 codec option - the output summary only ever lists IDR, I and P as encoded frame types. Not sure if this fact relates to another symptom; using VLC media player on a windows machine, the resulting output is not seek-able - in this particular case I can only ever seek back the very first frame of the video. Generally speaking, though, the files does play back well if one just lets it run from the beginning to the end. 

```
Input:   "D:\.Videos\Intermediate\test.mov"
Output:  "D:\.Videos\Rendered\test-av1-test.mkv"
Options: -c av1 --audio-copy
Press any key to continue . . .
D:\.Videos\NVenc\NVEncC64.exe -i "D:\.Videos\Intermediate\test.mov" -o "D:\.Videos\Rendered\test-av1-test.mkv" -c av1 --audio-copy
--------------------------------------------------------------------------------
D:\.Videos\Rendered\test-av1-test.mkv
--------------------------------------------------------------------------------
NVEncC (x64) 7.04 (r2374) by rigaya, Nov 12 2022 12:30:49 (VC 1929/Win)
OS Version     Windows 10 x64 (19044) [UTF-8]
CPU            AMD Ryzen 9 5950X 16-Core Processor [3.52GHz] (16C/32T)
GPU            #0: NVIDIA GeForce RTX 4090 (16384 cores, 2565 MHz)[PCIe4x16][522.25]
NVENC / CUDA   NVENC API 12.0, CUDA 11.8, schedule mode: auto
Input Buffers  CUDA, 16 frames
Input Info     avsw: dnxhd(yuv422)->nv12 [-], 3440x1440, 60/1 fps
AVSync         vfr
Vpp Filters    copyHtoD
Output Info    AV1 main @ Level auto
               3440x1440p 1:1 60.000fps (60/1fps)
               avwriter: av1, #1:pcm_s16le/stereo -> pcm_s16le/stereo/1536kbps
                => matroska
Encoder Preset default
Rate Control   CQP  I:20  P:23  B:25
ChromaQPOffset cb:0  cr:0
Lookahead      off
GOP length     600 frames
B frames       3 frames [ref mode: disabled]
Ref frames     3 frames, MultiRef L0:auto L1:auto
AQ             off
Part size      max auto / min auto
Tile num       columns auto / rows auto
TemporalLayers max 1
Refs           forward auto, backward auto
Others         mv:auto

encoded 34011 frames, 180.82 fps, 390332.56 kbps, 26376.25 MB
encode time 0:03:08, CPU: 25.3%, GPU: 72.7%, VE: 95.2%, GPUClock: 2079MHz, VEClock: 1799MHz
frame type IDR    57
frame type I      57,  avgQP  25.00,  total size     75.87 MB
frame type P   33954,  avgQP  24.49,  total size  26300.38 MB
Press any key to continue . . .
```

I was experimenting a bit by explicitly specifying --refs as well as --bframes, however, this did not result in any apparent B-frame-generation either; 

```
Input:   "D:\.Videos\Intermediate\test.mov"
Output:  "D:\.Videos\Rendered\test-av1-test.mkv"
Options: -c av1 --ref 16 --bframes 7 --audio-copy
Press any key to continue . . .
D:\.Videos\NVenc\NVEncC64.exe -i "D:\.Videos\Intermediate\test.mov" -o "D:\.Videos\Rendered\test-av1-test.mkv" -c av1 --ref 16 --bframes 7 --audio-copy
--------------------------------------------------------------------------------
D:\.Videos\Rendered\test-av1-test.mkv
--------------------------------------------------------------------------------
NVEncC (x64) 7.04 (r2374) by rigaya, Nov 12 2022 12:30:49 (VC 1929/Win)
OS Version     Windows 10 x64 (19044) [UTF-8]
CPU            AMD Ryzen 9 5950X 16-Core Processor [3.52GHz] (16C/32T)
GPU            #0: NVIDIA GeForce RTX 4090 (16384 cores, 2565 MHz)[PCIe4x16][522.25]
NVENC / CUDA   NVENC API 12.0, CUDA 11.8, schedule mode: auto
Input Buffers  CUDA, 20 frames
Input Info     avsw: dnxhd(yuv422)->nv12 [-], 3440x1440, 60/1 fps
AVSync         vfr
Vpp Filters    copyHtoD
Output Info    AV1 main @ Level auto
               3440x1440p 1:1 60.000fps (60/1fps)
               avwriter: av1, #1:pcm_s16le/stereo -> pcm_s16le/stereo/1536kbps
                => matroska
Encoder Preset default
Rate Control   CQP  I:20  P:23  B:25
ChromaQPOffset cb:0  cr:0
Lookahead      off
GOP length     600 frames
B frames       7 frames [ref mode: disabled]
Ref frames     16 frames, MultiRef L0:auto L1:auto
AQ             off
Part size      max auto / min auto
Tile num       columns auto / rows auto
TemporalLayers max 1
Refs           forward auto, backward auto
Others         mv:auto

encoded 34011 frames, 179.34 fps, 407044.21 kbps, 27505.52 MB
encode time 0:03:09, CPU: 25.0%, GPU: 76.9%, VE: 96.7%, GPUClock: 2118MHz, VEClock: 1811MHz
frame type IDR    57
frame type I      57,  avgQP  25.00,  total size     75.87 MB
frame type P   33954,  avgQP  24.74,  total size  27429.65 MB
Press any key to continue . . .
```

(Conversely, using -c hevc will confirm B-frame-generation in the output summary, for what it's worth) 

Am I missing something really obvious here?

**Comment by rigaya** [comment]
AV1 does have bi-directional predicted frames, but it is not like the way as H.264/HEVC. Instead it has so-called "mini-GOP", which includes series of inter frames that reference each other.

Because of this difference, it is not counted as B frames in the log.

I'm not sure why it isn't seekable, is it only with VLC, or the same with other players (like MPH-HC or MPC-BE)?

**Comment by IsmetBesic** [comment]
> AV1 does have bi-directional predicted frames, but it is not like the way as H.264/HEVC. Instead it has so-called "mini-GOP", which includes series of inter frames that reference each other.
> 
> Because of this difference, it is not counted as B frames in the log.

Thank you very much for the explanation! 

> I'm not sure why it isn't seekable, is it only with VLC, or the same with other players (like MPH-HC or MPC-BE)? 

Thanks for raising a couple of alternatives to try out - ultimately I did download MPH-HC (v1.9.24 - ffe50c0c7) and was able to replicate the same behavior. Even ended up trying out in chrome after encoding it to a .mp4 container - same behavior. 

However, I did eventually find the `Messages` window in VLC and was greeted by an endless logspam after seeking ahead; 

```
...
dav1d error: Decoder feed error -22!
dav1d error: Decoder feed error -22!
dav1d error: Decoder feed error -22!
...
```

Currently operating under the assumption that basically everyone is using dav1d as the underlying decoder implementation, hence why everything appears to be affected - though I would be curious to hear if anybody else is _not_ experiencing the same kind of playback issues.

**Comment by rigaya** [comment]
I've got RTX4080 for testing, and now I think I was able to fix non-seekable error in [NVEnc 7.05](https://github.com/rigaya/NVEnc/releases/tag/7.05).

Please have a try when you have time.

**Comment by IsmetBesic** [comment]
I can confirm that seeking works well when using version 7.05 to encode! Thanks a bunch!
